### PR TITLE
fix(dot/subscription): unsafe type casting from untrusted input

### LINKED ISSUE #2412: Unsafe Type Casting from Untrusted Input (GSR-14)
Websocket RPC messages are unmarshalled from JSON input before being processed.

After the JSON has been unmarshalled as a `map[string]interface{}` the fields are converted to their required types.

It is unsafe to assume data decoded from the websockets has the correct type as incorrect types will panic without using safe casting.

For example a malicious user could send a JSON message with `id` field as a `string` then `reqid := msg["id"].(float64)` will panic.

Consider doing safe, checked casting for all untrusted input. An example can be seen below.

Note: see GSR-15 for further additions to this code.

The `params` value also requires safe casting, however this must be done in each of the RPC methods individually when the type of `params` is known, such as `initExtrinsicWatch()`. 

```go
reqid, ok := msg["id"].(float64)
if !ok {
    // Handle Error
}
method, ok := msg["method"].(string)
if !ok {
    // Handle Error
}
```

### PR DESCRIPTION
**Pull request by EclesioMeloJunior**
## Changes

- Add checks in methods `initExtrinsicWatch` and `initStorageChangeListener` to safely casting types from untrusted inputs

## Tests

```sh
go test github.com/ChainSafe/gossamer/dot/rpc/subscription
```

## Issues

- Closes #2412

## Primary Reviewer

@edwardmack

### CODE REVIEW

File: dot/rpc/subscription/websocket.go [hunk 1]
```diff
@@ -26,8 +26,11 @@ type httpclient interface {
 	Do(*http.Request) (*http.Response, error)
 }
 
-var errCannotReadFromWebsocket = errors.New("cannot read message from websocket")
-var errCannotUnmarshalMessage = errors.New("cannot unmarshal webasocket message data")
+var (
+	errCannotReadFromWebsocket = errors.New("cannot read message from websocket")
+	errCannotUnmarshalMessage  = errors.New("cannot unmarshal webasocket message data")
```
Comment (qdm12) [inline dot/rpc/subscription/websocket.go:31]:
```suggestion
	errCannotUnmarshalMessage  = errors.New("cannot unmarshal websocket message data")
```

File: dot/rpc/subscription/websocket.go [hunk 2]
```diff
@@ -159,25 +162,24 @@ func (c *WSConn) initStorageChangeListener(reqID float64, params interface{}) (L
...
-			stgobs.filter[p] = []byte{}
-		default:
-			return nil, fmt.Errorf("unknown parameter type")
+
+			stgobs.filter[key] = []byte{}
 		}
+	case string:
+		stgobs.filter[filter] = []byte{}
+	default:
+		return nil, fmt.Errorf("expected type equals string or []string. got %T", params)
```
Comment (qdm12) [inline dot/rpc/subscription/websocket.go:182]: use a sentinel error

File: dot/rpc/subscription/websocket.go [hunk 3]
```diff
@@ -159,25 +162,24 @@ func (c *WSConn) initStorageChangeListener(reqID float64, params interface{}) (L
...
-					return nil, fmt.Errorf("unknown parameter type")
-				}
-				stgobs.filter[data] = []byte{}
+	switch filter := params.(type) {
+	case []string:
+		for _, key := range filter {
+			stgobs.filter[key] = []byte{}
+		}
+	case []interface{}:
+		for _, interfaceKey := range params.([]interface{}) {
```
Comment (qdm12) [inline dot/rpc/subscription/websocket.go:171]: I think that would work right?

```suggestion
		for _, interfaceKey := range filter {
```

File: dot/rpc/subscription/websocket.go [hunk 4]
```diff
@@ -265,14 +267,30 @@ func (c *WSConn) initAllBlocksListerner(reqID float64, _ interface{}) (Listener,
...
+		encodedExtrinsic = encodedHex[0]
+	case []interface{}:
+		if len(encodedHex) != 1 {
+			return nil, fmt.Errorf("expected only 1 param item. got %d", len(encodedHex))
+		}
+
+		var ok bool
+		encodedExtrinsic, ok = encodedHex[0].(string)
+		if !ok {
+			return nil, fmt.Errorf("expected type equals string. got %T", encodedHex[0])
```
Comment (qdm12) [inline dot/rpc/subscription/websocket.go:286]: use a sentinel error

File: dot/rpc/subscription/websocket.go [hunk 5]
```diff
@@ -265,14 +267,30 @@ func (c *WSConn) initAllBlocksListerner(reqID float64, _ interface{}) (Listener,
...
+		}
+	default:
+		return nil, fmt.Errorf("expected type equals []string. got %T", params)
```
Comment (qdm12) [inline dot/rpc/subscription/websocket.go:289]: use a sentinel error

also that should be `[]string or []interface{}` I guess? :thinking:

File: dot/rpc/subscription/websocket.go [hunk 6]
```diff
@@ -159,25 +162,24 @@ func (c *WSConn) initStorageChangeListener(reqID float64, params interface{}) (L
...
-				if !ok {
-					return nil, fmt.Errorf("unknown parameter type")
-				}
-				stgobs.filter[data] = []byte{}
+	switch filter := params.(type) {
+	case []string:
+		for _, key := range filter {
+			stgobs.filter[key] = []byte{}
+		}
+	case []interface{}:
```
[the remark this answers is not in the document]
Comment (EclesioMeloJunior) [reply dot/rpc/subscription/websocket.go:170]: > Couldn't a string be interpreted as []interface{} (slice of runes)?

Not really -> https://go.dev/play/p/cCeKJJsIBU8

[the remark this answers is not in the document]
Comment (EclesioMeloJunior) [reply dot/rpc/subscription/websocket.go:174]: the method `initStorageChangeListener(reqID float64, params interface{})` receives the `params` which is a dynamic websocket data like:  multi-type slice`["hash", 10]`, a hex string or just a number, so adding this case I make sure that we cover the type`[]interface{"a", "c"}` as well as `[]string{"a", "c"}`

File: dot/rpc/subscription/websocket.go (hunk 5 above)
Comment (EclesioMeloJunior) [reply dot/rpc/subscription/websocket.go]: The method `initExtrinsicWatch(reqID float64, params interface{})` expects a slice of 1 item that must be a string so there is the `case []interface{}` just to cover the case of `[]interface{"a"}` but the data should be always []string of 1 item

File: dot/rpc/subscription/websocket.go [hunk 7]
```diff
@@ -265,14 +267,30 @@ func (c *WSConn) initAllBlocksListerner(reqID float64, _ interface{}) (Listener,
...
 
-	if len(pA) != 1 {
-		return nil, errors.New("expecting only one parameter")
+	switch encodedHex := params.(type) {
+	case []string:
+		if len(encodedHex) != 1 {
+			return nil, fmt.Errorf("expected only 1 param item. got %d", len(encodedHex))
+		}
+		encodedExtrinsic = encodedHex[0]
+	case []interface{}:
```
[the remark this answers is not in the document]
Comment (EclesioMeloJunior) [reply dot/rpc/subscription/websocket.go:280]: https://github.com/ChainSafe/gossamer/pull/2529#discussion_r872648722

File: dot/rpc/subscription/websocket.go [hunk 8]
```diff
@@ -265,14 +267,30 @@ func (c *WSConn) initAllBlocksListerner(reqID float64, _ interface{}) (Listener,
 }
 
 func (c *WSConn) initExtrinsicWatch(reqID float64, params interface{}) (Listener, error) {
```
[the remark this answers is not in the document]
Comment (EclesioMeloJunior) [reply dot/rpc/subscription/websocket.go:271]: No, this param should be an `interface{}` as it is what we receive from websocket, so each function is responsible to cast the data in the format it expects

File: dot/rpc/subscription/websocket.go [hunk 9]
```diff
@@ -159,25 +164,24 @@ func (c *WSConn) initStorageChangeListener(reqID float64, params interface{}) (L
...
-		case []interface{}:
-			for _, pp := range param.([]interface{}) {
-				data, ok := pp.(string)
-				if !ok {
-					return nil, fmt.Errorf("unknown parameter type")
-				}
-				stgobs.filter[data] = []byte{}
+	switch filter := params.(type) {
+	case string:
+		stgobs.filter[filter] = []byte{}
```
Comment (qdm12) [inline dot/rpc/subscription/websocket.go:169]: why do we set this to an empty slice and not `nil`? Setting it to `nil` would still set the key in the map, and we can then set it or append to it as well later.

File: dot/rpc/subscription/websocket.go (hunk 5 above)
Comment (qdm12) [reply dot/rpc/subscription/websocket.go]: Oh yes I was talking about the error string here

File: dot/rpc/subscription/websocket.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (qdm12) [reply dot/rpc/subscription/websocket.go:174]: Awesome TIL!

File: dot/rpc/subscription/websocket.go (hunk 5 above)
Comment (qdm12) [reply dot/rpc/subscription/websocket.go]: I was talking about the error string. I made a code suggestion below 😉

File: dot/rpc/subscription/websocket_test.go [hunk 10]
```diff
@@ -207,7 +206,7 @@ func TestWSConn_HandleComm(t *testing.T) {
 	wsconn.CoreAPI = modules.NewMockCoreAPI()
 	wsconn.BlockAPI = nil
 	wsconn.TxStateAPI = modules.NewMockTransactionStateAPI()
-	listner, err := wsconn.initExtrinsicWatch(0, []interface{}{"NotHex"})
+	listner, err := wsconn.initExtrinsicWatch(0, []string{"NotHex"})
```
[the remark this answers is not in the document]
Comment (EclesioMeloJunior) [reply dot/rpc/subscription/websocket_test.go:209]: for `initExtrinsicWatch` the cases: `[]string` and `[]interface` being tested and for `initStorageChangeListener` the cases: `string`, `[]string` and `[]interface` are being tested as well, I've created [an issue](https://github.com/ChainSafe/gossamer/issues/2555) to refactor those tests to support test cases and then they will be more readable

File: dot/rpc/subscription/websocket.go (hunk 9 above)
Comment (qdm12) [reply dot/rpc/subscription/websocket.go:169]: Just a few bytes less. I guess not a big deal either, so up to you!

File: dot/rpc/subscription/websocket.go [hunk 11]
```diff
@@ -265,14 +271,32 @@ func (c *WSConn) initAllBlocksListerner(reqID float64, _ interface{}) (Listener,
...
 func (c *WSConn) initExtrinsicWatch(reqID float64, params interface{}) (Listener, error) {
-	pA := params.([]interface{})
+	var encodedExtrinsic string
 
-	if len(pA) != 1 {
-		return nil, errors.New("expecting only one parameter")
+	switch encodedHex := params.(type) {
+	case []string:
+		if len(encodedHex) != 1 {
+			return nil, fmt.Errorf("%w: %d, expected 1 param", errUnexpectedParamLen, len(encodedHex))
```
Comment (jimjbrettj) [inline dot/rpc/subscription/websocket.go:279]: Nit: Should we be more clear on what we got vs what was expected? Something like fmt.Errorf("%w: expected 1 param, got %d", errUnexpectedParamLen, len(encodedHex)).

**Pull request merged**